packages/language-server/lib/convert.js

```diff
@@ -0,0 +1,365 @@
+ * @typedef {import('typescript').DiagnosticRelatedInformation} DiagnosticRelatedInformation
+ * @typedef {import('typescript').JSDocTagInfo} JSDocTagInfo
+ * @typedef {import('typescript').NavigationBarItem} NavigationBarItem
+ * @typedef {import('typescript').QuickInfo} QuickInfo
+ * @typedef {import('typescript').SymbolDisplayPart} SymbolDisplayPart
+ * @typedef {import('typescript').ScriptElementKind} ScriptElementKind
+ * @typedef {import('typescript').TextSpan} TextSpan
+ * @typedef {import('vscode-languageserver').Diagnostic} LspDiagnostic
+ * @typedef {import('vscode-languageserver').DiagnosticRelatedInformation} LspDiagnosticRelatedInformation
+ * @typedef {import('vscode-languageserver-textdocument').TextDocument} TextDocument
+ */
+
+import {pathToFileURL} from 'node:url'
+
+import {
+  CompletionItemKind,
+  DiagnosticSeverity,
+  DiagnosticTag,
+  DocumentSymbol,
+  LocationLink,
+  Range,
+  SymbolKind
+} from 'vscode-languageserver'
+
+import {documents} from './documents.js'
+
+/**
+ * @param {ScriptElementKind} kind
+ * @returns {CompletionItemKind} The matching Monaco completion item kind.
+ */
+export function convertScriptElementKind(kind) {
+  switch (kind) {
+    case 'primitive type':
+    case 'keyword': {
+      return CompletionItemKind.Keyword
+    }
+
+    case 'var':
+    case 'local var': {
+      return CompletionItemKind.Variable
+    }
+
+    case 'property':
+    case 'getter':
+    case 'setter': {
+      return CompletionItemKind.Field
+    }
+
+    case 'function':
+    case 'method':
+    case 'construct':
+    case 'call':
+    case 'index': {
+      return CompletionItemKind.Function
+    }
+
+    case 'enum': {
+      return CompletionItemKind.Enum
+    }
+
+    case 'module': {
+      return CompletionItemKind.Module
+    }
+
+    case 'class': {
+      return CompletionItemKind.Class
+    }
+
+    case 'interface': {
+      return CompletionItemKind.Interface
+    }
+
+    case 'warning': {
+      return CompletionItemKind.File
+    }
+
+    default: {
+      return CompletionItemKind.Property
+    }
+  }
+}
+
+/**
+ * @param {ScriptElementKind} kind
+ * @returns {SymbolKind} The matching Monaco completion item kind.
+ */
+export function convertScriptElementKindToSymbolKind(kind) {
+  switch (kind) {
+    case 'property':
+    case 'getter':
+    case 'setter': {
+      return SymbolKind.Property
+    }
+
+    case 'function':
+    case 'method':
+    case 'construct':
+    case 'call':
+    case 'index': {
+      return SymbolKind.Function
+    }
+
+    case 'enum': {
+      return SymbolKind.Enum
+    }
+
+    case 'module': {
+      return SymbolKind.Module
+    }
+
+    case 'class': {
+      return SymbolKind.Class
+    }
+
+    case 'interface': {
+      return SymbolKind.Interface
+    }
+
+    default: {
+      return SymbolKind.Variable
+    }
+  }
+}
+
+/**
+ * @param {ts} ts
+ * @param {CompletionEntryDetails | QuickInfo} details
+ * @returns {string} XXX
```

Comment:
Should these placeholders now we solved?

Reply:
Yes. Should I just remove them? I added them, because the previous ESLint required them. I believe most internals in the unified ecosystem relies on inferred return types and often leaves internal functions undocumented.

Reply:
Sometimes inferring return values is fine.
Inferring is nice, but at some point it’s good to define things, so that you know the inferring in other places is going okay.
So generally I like functions as boundaries for that currently: to define types of params and return values explicitly.

Whether you need to document everything: on the public API, yes. Internal, it depends.

Reply:
I decided to just document everything. :)